src/botnet.c

```diff
@@ -1468,36 +1468,45 @@ static void eof_dcc_relaying(int idx)
 
 static void dcc_relay(int idx, char *buf, int j)
 {
-  unsigned char *p = (unsigned char *) buf;
-  int mark;
+  unsigned char *src, *dst;
 
   for (j = 0; dcc[j].sock != dcc[idx].u.relay->sock ||
        dcc[j].type != &DCC_RELAYING; j++);
-  /* If redirecting to a non-telnet user, swallow telnet codes and
-   * escape sequences. */
+  /* If redirecting to a non-telnet user, swallow telnet IAC, escape sequences
+   * and CR. */
   if (!(dcc[j].status & STAT_TELNET)) {
-    while (*p != 0) {
-      while (*p != 255 && (*p != '\033' || *(p + 1) != '[') && *p != '\r' && *p)
-        p++;                    /* Search for IAC, escape sequences and CR. */
-      if (*p == 255) {
-        mark = 2;
-        if (!*(p + 1))
-          mark = 1;             /* Bogus */
-        if ((*(p + 1) >= 251) || (*(p + 1) <= 254)) {
-          mark = 3;
-          if (!*(p + 2))
-            mark = 2;           /* Bogus */
+    src = (unsigned char *) buf;
+    dst = (unsigned char *) buf;
+    while (*src) {
+      /* Search for IAC, escape sequences and CR. */
+      if (*src == TLN_IAC) {
+        src++;
+        if ((*src >= TLN_WILL) && (*src <= TLN_DONT)) {
+          src++;
+          if (*src)
+            src++;
         }
-        strcpy((char *) p, (char *) (p + mark));
-      } else if (*p == '\033') {
-        unsigned char *e;
-
-        /* Search for the end of the escape sequence. */
-        for (e = p + 2; *e != 'm' && *e; e++);
-        strcpy((char *) p, (char *) (e + 1));
-      } else if (*p == '\r')
-        memmove(p, p + 1, strlen((char *)p + 1) + 1);
+        else if (*src)
+          src++;
+      } else if (*src == '\033') { /* ESC */
+        src++;
+        if (*src == '[') { /* CSI */
+          src++;
+          /* Search for the end of the escape sequence. */
+          while (*src && *src != 'm')
+            src++;
+        }
```

Comment:
If I'm not mistaken, the `'m'` is not skipped and will, in the next loop, be added to dst.